client-app/tests/unit/query-builder.spec.ts

```diff
@@ -0,0 +1,102 @@
+    const query = keywordQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("keyword=test");
+  });
+});
+
+describe("terms: parse and build", () => {
+  const productQueryBuilder = new QueryBuilder(ProductSearchCriteria, ProductSearchQuery);
+  const query = new URLSearchParams("sort_by=createddate-descending&terms=Brand%3a3DR%2cDJI%3bprice%3aover-1000");
+
+  it("should correctly parse sorting and terms", () => {
+    const searchCriteria = productQueryBuilder.fromQuery(query);
+    expect(searchCriteria.sortBy).to.be.equal("createddate-descending");
+    expect(searchCriteria.terms?.items).to.have.property("Brand");
+    expect(searchCriteria.terms?.items["Brand"]).to.include("3DR");
+    expect(searchCriteria.terms?.items["Brand"]).to.include("DJI");
+    expect(searchCriteria.terms?.items).to.have.property("price");
+    expect(searchCriteria.terms?.items["price"]).to.include("over-1000");
+  });
+
+  it("should correctly build query with sorting and terms", () => {
+    const searchCriteria = new ProductSearchCriteria(undefined, undefined, undefined, "createddate-descending");
+    searchCriteria.terms = new Terms();
+    searchCriteria.terms.items["Brand"] = ["3DR", "DJI"];
+    searchCriteria.terms.items["price"] = ["over-1000"];
+    const query = productQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("sort_by=createddate-descending&terms=Brand%3a3DR%2cDJI%3bprice%3aover-1000");
+  });
+});
+
+describe("terms: toggle", () => {
+  const productQueryBuilder = new QueryBuilder(ProductSearchCriteria, ProductSearchQuery);
+  const searchCriteria = new ProductSearchCriteria();
+  searchCriteria.terms = new Terms();
+  searchCriteria.terms.items["Brand"] = ["3DR", "DJI"];
+  searchCriteria.terms.items["price"] = ["over-1000"];
+
+  it("should remove existing value from terms", () => {
+    searchCriteria.terms?.toggle("Brand", "3DR");
+    const query = productQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("terms=Brand%3aDJI%3bprice%3aover-1000");
+  });
+
+  it("should remove key from terms", () => {
+    searchCriteria.terms?.toggle("price", "over-1000");
+    const query = productQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("terms=Brand%3aDJI");
+  });
+
+  it("should add new key and value to terms", () => {
+    // Add new value
+    searchCriteria.terms?.toggle("price", "600-1000");
+    const query = productQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("terms=Brand%3aDJI%3bprice%3a600-1000");
+  });
+
+  it("should add value to existing key in terms", () => {
+    // Add value to existing key
+    searchCriteria.terms?.toggle("price", "over-1000");
+    const query = productQueryBuilder.toQuery(searchCriteria);
+    expect(query.toString()).to.be.equalIgnoreCase("terms=Brand%3aDJI%3bprice%3a600-1000%2cover-1000");
+  });
```

Comment:
It's a pleasure to see tests here!
I guess we can use here some ideas from this documentation https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices

eg. 
- test organization (AAA approach)
- not test many things in one test
etc

Reply:
1. I'm trying to follow AAA approach. I'm a little bit refactored & formatted code tow show that every test has 3 parts (in terms I do arrange in suite)
2. I need to test correctness of parsed params and terms. That's why it have so much assertions

Please review again with changes